spice/spice.py

```diff
@@ -129,110 +156,87 @@ def __init__(self, model=None, provider=None, model_aliases=None):
                 )
 
-    async def call_llm(
-        self,
-        messages: list[dict[str, str]],
-        model: Optional[str] = None,
-        stream: bool = False,
-        temperature: Optional[float] = None,
-        max_tokens: Optional[int] = None,
-        response_format: Optional[dict[str, str]] = None,
-        logging_callback: Optional[Callable[[SpiceResponse], None]] = None,
-    ) -> SpiceResponse:
+    def _get_model(self, model: Optional[str]):
         if model is None:
             if self._default_model is None:
-                raise SpiceError("model argument is required when default model is not set at initialization")
+                raise SpiceError("Model argument is required when default model is not set at initialization")
             model = self._default_model
 
         if self._model_aliases is not None:
             if model in self._model_aliases:
                 model = self._model_aliases[model]["model"]
             else:
                 raise SpiceError(f"Unknown model alias: {model}")
+        return model
 
-        model = cast(str, model)
-
+    def _get_client(self, model: str):
         if self._default_client is not None:
-            client = self._default_client
+            return self._default_client
         else:
             provider = _get_provider_from_model_name(model)
             if provider not in self._clients:
-                raise SpiceError(f"Provider {provider} is not set up for model {model}")
-            client = self._clients[provider]
+                raise SpiceError(f"Provider {provider} is not currently supported.")
+            return self._clients[provider]
 
-        # not all providers support response format
+    def _fix_call_args(
+        self,
+        messages: List[SpiceMessage],
+        model: str,
+        stream: bool,
+        temperature: Optional[float],
+        max_tokens: Optional[int],
+        response_format: Optional[ResponseFormatType],
+    ):
+        # Not all providers support response format
         if response_format is not None:
             if response_format == {"type": "text"}:
                 response_format = None
 
-        # gpt-4-vision has low default max_tokens
-        if max_tokens is None and model == "gpt-4-vision-preview":
-            max_tokens = 4096
-
-        response = SpiceResponse(
-            call_args=SpiceCallArgs(
-                model=model,
-                messages=messages,
-                stream=stream,
-                temperature=temperature,
-                max_tokens=max_tokens,
-                response_format=response_format,
-            ),
-            logging_callback=logging_callback,
+        return SpiceCallArgs(
+            model=model,
+            messages=messages,
+            stream=stream,
+            temperature=temperature,
+            max_tokens=max_tokens,
+            response_format=response_format,
         )
 
+    async def completion(
```

Comment:
at first I didn't like using the names `completion`/`stream_completion` because it sounds like the older completions api that OpenAI and Anthropic had (gpt 3 davinci, other models) before they updated to Chat Completions and Messages. But maybe it's fine? `call_llm` is a bit wrong as well since the models are multimodal. `call_model` would be nice except it doesn't make it clear it's not an embedding or whisper call.

Reply:
I personally like completion but if you have an idea you like more I'd be willing to change.